Event formatter batch handling

jskeet opened the pull request:
(This comment is now out of date; see comment from 2021-04-08 for update.)

First two commits are from #110.

Obviously we'd need tests, and then probably protocol binding support (at least for HTTP). This is mostly to show what the CloudEventFormatter interface might look like for batch.

I'm nervous about it because it's not a separate content mode in the CloudEvents spec - but it would be really hard to express without putting it into CloudEventFormatter. It probably doesn't make sense to include it in the ContentMode enum, because the signatures for protocol bindings don't make sense (as they need to accept/return a collection instead of a single event). Basically it doesn't fit in terribly neatly...

JoshLove-msft commented on src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs:
```diff
@@ -120,31 +120,72 @@ public JsonEventFormatter(JsonSerializerOptions serializerOptions, JsonDocumentO
+        private CloudEvent DecodeJsonElement(JsonElement element, IEnumerable<CloudEventAttribute> extensionAttributes = null)
         {
-            if (!document.RootElement.TryGetProperty(CloudEventsSpecVersion.SpecVersionAttribute.Name, out var specVersionProperty)
+            if (element.ValueKind != JsonValueKind.Object)
```
What about a single element array?

jskeet replied:
I personally think trying to parse a single-element batch as if it's not a batch at all should still fail. The calling code is doing the wrong thing, and that should be reported as early as possible - otherwise I can see code that works in all tests and then fails with real batches. Happy to be persuaded though.

JoshLove-msft replied:
I was thinking of the case where a caller might know that they will always get a single event in a JSON array. For instance, maybe the publisher always publishes it this way. But I think you raise a fair point about the potential for missing this during testing.

jskeet replied:
Yup - will leave it for now. I suspect we may want more introspection capabilities on event formatters before we go GA, so you could have multiple formatters and pick the right one to use (and the right method to use, potentially) based on the content type.

jskeet commented:
Will try to turn this into a non-prototype later this week.

JoshLove-msft commented on src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs:
```diff
@@ -136,6 +136,44 @@ public override CloudEvent DecodeStructuredModeMessage(Stream data, ContentType
+            return DecodeJArray(array, extensionAttributes);
+        }
+
+        public override IReadOnlyList<CloudEvent> DecodeBatchModeMessage(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
```
Do we know if there will ever be a Binary batch message as part of the spec?

jskeet replied:
I've seen no hint of that, no. I can't see how it would work to be honest, given that headers are used for the attributes, and each event would need its own set of attributes. (I'm sure it could be done, but I suspect it would be massively clunky compared with just using the current approach.) I think we're okay to design the SDK (and method names in particular) assuming that batch always means "in a structured mode style".

JoshLove-msft commented on src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs:
```diff
@@ -136,6 +136,44 @@ public override CloudEvent DecodeStructuredModeMessage(Stream data, ContentType
         public override CloudEvent DecodeStructuredModeMessage(byte[] data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
             DecodeStructuredModeMessage(new MemoryStream(data), contentType, extensionAttributes);
 
```
Update to mention new method.

JoshLove-msft commented on src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs:
```diff
@@ -136,6 +136,44 @@ public override CloudEvent DecodeStructuredModeMessage(Stream data, ContentType
         public override CloudEvent DecodeStructuredModeMessage(byte[] data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
             DecodeStructuredModeMessage(new MemoryStream(data), contentType, extensionAttributes);
 
+        public override async Task<IReadOnlyList<CloudEvent>> DecodeBatchModeMessageAsync(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
```
Consider updating to mention new methods https://github.com/cloudevents/sdk-csharp/pull/111/files#diff-b73904f7e0560631ddf370286e4d19cd3edf58de51d76bc28a6d33b11edb5b94R59

jskeet replied:
Yes indeed! Thanks for the reminder. Will definitely do that.

JoshLove-msft commented on src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs:
```diff
@@ -136,6 +136,44 @@ public override CloudEvent DecodeStructuredModeMessage(Stream data, ContentType
+
+        private IReadOnlyList<CloudEvent> DecodeJArray(JArray jArray, IEnumerable<CloudEventAttribute> extensionAttributes = null)
+        {
+            List<CloudEvent> events = new List<CloudEvent>();
```
Should we use jArray.Count when initializing the list?

jskeet replied:
It certainly wouldn't hurt. (Alternatively this method could be rewritten via LINQ... I might have a try at that, as it feels odd to use a loop here.)

jskeet replied:
Done - decided not to use LINQ for the same reasons as in the System.Text.Json code; it seems unnecessarily inefficient and wouldn't actually be much simpler.

JoshLove-msft commented on src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs:
```diff
@@ -136,6 +136,44 @@ public override CloudEvent DecodeStructuredModeMessage(Stream data, ContentType
+                }
+                else
+                {
+                    throw new ArgumentException("Invalid array element within batch", "data");
```
consider passing in paramName

JoshLove-msft replied:
Should we include the token in the exception?

jskeet replied:
Happy to pass in paramName to avoid the hard-coding, yes - thanks.
I don't want to include the actual token, as generally I don't like exceptions containing potentially arbitrary data (both in terms of size and PII)... but we could include the type.

JoshLove-msft replied:
Right.. I forgot about that regarding including arbitrary data. We can also include the index of the failing element.

jskeet replied:
Done.

JoshLove-msft commented on src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs:
```diff
@@ -306,16 +344,43 @@ protected virtual void DecodeStructuredModeDataBase64Property(JToken dataBase64T
+        public override byte[] EncodeBatchModeMessage(IEnumerable<CloudEvent> cloudEvents, out ContentType contentType)
+        {
+            cloudEvents = cloudEvents ?? throw new ArgumentNullException(nameof(cloudEvents));
+            contentType = new ContentType("application/cloudevents-batch+json")
```
Consider putting this in a constants file along with the content type for structured mode.

jskeet replied:
Definitely happy to make it a constant... not sure about a file just for constants though. I could put the prefix in MimeUtilities which is where most of our content type common code is. Will have a look.

JoshLove-msft commented on src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs:
```diff
@@ -120,31 +120,72 @@ public JsonEventFormatter(JsonSerializerOptions serializerOptions, JsonDocumentO
-            if (!document.RootElement.TryGetProperty(CloudEventsSpecVersion.SpecVersionAttribute.Name, out var specVersionProperty)
+            if (element.ValueKind != JsonValueKind.Object)
+            {
+                throw new ArgumentException($"Cannot decode JSON element of kind '{element.ValueKind}' as CloudEvent");
```
Might be worth including the element itself as well.

jskeet replied:
As per earlier comment, I'd generally prefer to avoid including arbitrary data in the exception message. I could be persuaded though - let's chat about it.

jskeet reviewed (commented):
Thanks for the comments. I'll *try* to come back to this PR before the end of the week - adding tests as well as implementing your comments. I may not get to it though.

jskeet commented:
I've now rebased, addressed review comments, and added a commit with tests.

@JoshLove-msft please re-review with those changes... if you're happy, I'll merge and we can then get to the protocol binding methods.

jskeet commented:
(Rebased back to two commits, ready to merge on green.)
